js/app/modules/selection/xapian.js

```diff
@@ -85,24 +85,43 @@ const Xapian = new Module.Class({
                 return;
             }
 
-            let next_query = info.more_results;
-            if (!next_query) {
+            let results_added = results.filter(this.add_model, this);
+            let need_more_results = num_desired > results_added.length;
+
+            let more_results = info.more_results;
+            if (!!more_results) {
+                // If we got `more_results`, check to see if we're at the limit
+                let can_load_more = info.upper_bound > results.length;
+                if (can_load_more !== this._can_load_more) {
+                    this._can_load_more = can_load_more;
+                    this.notify('can-load-more');
+                }
+
+                // Only recurse if we need more results and can get more
+                if (need_more_results && this._can_load_more) {
+                    this._next_query = more_results
+                    this.queue_load_more(num_desired - results_added.length);
+                }
+            } else {
+                // No `more_results` - see if we have additional queries to fill
+                // the results
                 this._query_index++;
-                next_query = this.construct_query_object(num_desired, this._query_index);
-            }
+                this._next_query = this.construct_query_object(num_desired,
+                                                               this._query_index);
+
+                // We have a new query. We only assume that we can get more results
+                // if we filled up the desired count. If the next query isn't defined,
+                // we know for sure that we won't be able to fill the results array.
+                let can_load_more = !need_more_results || !!this._next_query;
```

Comment:
@ptomato Hmm, I see your point now but effectively `need_more_results` == 'is query exhausted'. Ideally the model only takes the exact number of articles it originally asked for with `this.add_model` so the same query _may_ have results that weren't added the first time around so we do a slight assumption that if we filled up the needed results, there might be more results and if we didn't fill up the results we know for sure that we can't load more.